File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-2.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        maxArea = 0
        indexes = []
        for i in range(len(heights)):
            while len(indexes) > 0 and heights[indexes[-1]] > heights[i]:
                minIdx = indexes[-1]
                indexes.pop()

                width = i if len(indexes) == 0 else i - indexes[-1] - 1
                maxArea = max(maxArea, heights[minIdx] * width)
            indexes.append(i)
        
        while len(indexes) > 0:
            minIdx = indexes[-1]
            indexes.pop()

            width = len(heights) if len(indexes) == 0 else len(heights) - indexes[-1] - 1
            maxArea = max(maxArea, heights[minIdx] * width)
        return maxArea

File: Data Structures & Algorithms/max-water-container/submission-2.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {
        int i = 0;
        int j = heights.size() - 1;
        int maxA = 0;
        while(i < j){
            int minH = min(heights[i], heights[j]);
            int currA = minH * (j - i);
            maxA = max(maxA, currA);
            if(heights[i] < heights[j]) i++;
            else j--;
        }
        return maxA;
    }
};
